### PR TITLE
Added pre-trained GNN embeddings

### PR DESCRIPTION
**Pull request by leojklarner**
I refactored the code from the STRATEGIES FOR PRE-TRAINING GRAPH NEURAL NETWORKS paper for the generation of node embeddings from pre-trained parameters.
I included tests to make sure everything works and showcase how to use the network to derive the embeddings.
The failed CI checks come from the dataloader scripts, which have to be re-written at some point to fit the additional kernels we provide.
I didn't implement a graph-level aggregation step yet, as I thought that could depend on what downstream kernel we choose.

### CODE REVIEW

File: gprotorch/kernels/gnn_kernels/test_hu_iclr_2020.py [hunk 1]
```diff
@@ -0,0 +1,357 @@
+"""
+Reformatted and cleaned up GNN modules from Hu et al.
+Strategies for Pre-training Graph Neural Networks. ICLR 2020
+(https://github.com/snap-stanford/pretrain-gnns)
+
+Author: Leo Klarner (https://github.com/leojklarner), April 2022
+"""
+
```
Comment (Ryan-Rhys) [inline gprotorch/kernels/gnn_kernels/test_hu_iclr_2020.py:8]: Should we call the module test if it's not in the tests folder?

File: gprotorch/kernels/gnn_kernels/test_hu_iclr_2020.py [hunk 2]
```diff
@@ -0,0 +1,357 @@
+"""
+Reformatted and cleaned up GNN modules from Hu et al.
+Strategies for Pre-training Graph Neural Networks. ICLR 2020
+(https://github.com/snap-stanford/pretrain-gnns)
```
Comment (Ryan-Rhys) [inline gprotorch/kernels/gnn_kernels/test_hu_iclr_2020.py:4]: Should we also give the module this code is based on i.e. model.py in the linked repo?

File: gprotorch/kernels/gnn_kernels/test_hu_iclr_2020.py [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+            # x are atom features of the molecule and edge_attr the atomic features of the molecule
+            x = self.gnns[layer](x, edge_index, edge_attr)
+            x = self.batch_norms[layer](x)
+            if layer != self.num_layers - 1:
+                x = F.relu(x)
+
+        return x
+
+
+if __name__ == "__main__":
```
Comment (Ryan-Rhys) [inline gprotorch/kernels/gnn_kernels/test_hu_iclr_2020.py:356]: Is this necessary? (I know it's included in the original repo, I was just wondering!)

File: gprotorch/kernels/gnn_kernels/test_hu_iclr_2020.py [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+import numpy as np
+import torch
+import torch.nn.functional as F
+from rdkit import Chem
+from torch_geometric.data import Data
+from torch_geometric.nn import MessagePassing
+from torch_geometric.utils import add_self_loops
+from torch_scatter import scatter_add
+
+# define the allowable node and edge labels as used in Hu et al.
```
Comment (Ryan-Rhys) [inline gprotorch/kernels/gnn_kernels/test_hu_iclr_2020.py:20]: Are these standard settings for running GNNs on molecules?

File: gprotorch/kernels/gnn_kernels/test_hu_iclr_2020.py [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+    is compatible with the GNN modules below.
+
+    Args:
+        mol: rdkit mol object
+
+    Returns: PyTorch Geometric data object
+
+    """
+
+    # derive atom features: atomic number + chirality tag
```
Comment (Ryan-Rhys) [inline gprotorch/kernels/gnn_kernels/test_hu_iclr_2020.py:68]: What does the chirality tag look like out of interest?

Comment (Ryan-Rhys) [reply gprotorch/kernels/gnn_kernels/test_hu_iclr_2020.py:68]: Can it help to enrich the atom features with other properties?

File: gprotorch/kernels/gnn_kernels/test_hu_iclr_2020.py [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+    Extension of the Graph Isomorphism Network to incorporate
+    edge information by concatenating edge embeddings.
+    """
+
+    def __init__(self, emb_dim, aggr="add"):
+        """
+        Initialise GIN convolutional layer.
+        Args:
+            emb_dim: latent node embedding dimension
+            aggr: aggregation procedure
```
Comment (Ryan-Rhys) [inline gprotorch/kernels/gnn_kernels/test_hu_iclr_2020.py:122]: What are the different options for the aggregation procedure out of interest?

File: tests/kernels/gnn_kernels/test_hu_iclr_2020.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+Hu et al. Strategies for Pre-training Graph Neural Networks. ICLR 2020
+(https://github.com/snap-stanford/pretrain-gnns)
+runs and produces node embeddings with the provided pre-trained parameters.
+"""
+
+import os
+
+import pytest
+import torch
+from gprotorch.dataloader import DataLoaderMP
```
Comment (Ryan-Rhys) [inline tests/kernels/gnn_kernels/test_hu_iclr_2020.py:12]: I think the "optimise imports" command in PyCharm (can't remember which dropdown menu it's under) will re-order these automatically so we have 1. python modules, 2. third-party modules and 3. our own modules in separate blocks in line with PEP8.

File: gprotorch/kernels/gnn_kernels/test_hu_iclr_2020.py (hunk 1 above)
Comment (leojklarner) [reply gprotorch/kernels/gnn_kernels/test_hu_iclr_2020.py]: Thanks, my bad!

File: gprotorch/kernels/gnn_kernels/test_hu_iclr_2020.py (hunk 2 above)
Comment (leojklarner) [reply gprotorch/kernels/gnn_kernels/test_hu_iclr_2020.py]: Good point.

File: gprotorch/kernels/gnn_kernels/test_hu_iclr_2020.py (hunk 3 above)
Comment (leojklarner) [reply gprotorch/kernels/gnn_kernels/test_hu_iclr_2020.py]: When I wrote this I did some quick testing directly from this script, but as we don't intend to run it as main we can take the main guard out.

File: gprotorch/kernels/gnn_kernels/test_hu_iclr_2020.py (hunk 5 above)
Comment (leojklarner) [reply gprotorch/kernels/gnn_kernels/test_hu_iclr_2020.py]: For atoms with four unique neighbours it's whether they are ordered clockwise (- form) or counter-clockwise (+ form), see https://en.wikipedia.org/wiki/Chirality_(chemistry).
Yes, including additional atom features would probably lead to improved performance, but you'd have to retrain the self-supervised initialisations.

File: gprotorch/kernels/gnn_kernels/test_hu_iclr_2020.py (hunk 6 above)
Comment (leojklarner) [reply gprotorch/kernels/gnn_kernels/test_hu_iclr_2020.py]: For GINs you can also do mean pooling and element-wise max pooling.

File: tests/kernels/gnn_kernels/test_hu_iclr_2020.py (hunk 7 above)
Comment (leojklarner) [reply tests/kernels/gnn_kernels/test_hu_iclr_2020.py:12]: I formatted them with isort. But should probably run the entire codebase through a reformatter anyway, once we're there.

**Pull request merged**